services/outputhost/replicaconnection.go

```diff
@@ -186,124 +188,176 @@ func (conn *replicaConnection) readMessagesPump() {
+				msgSeqNum = 0
+
+			} else if lastSeqNum+1 != int64(msgSeqNum) {
+
+				// FIXME: add metric to help alert this case
+				expectedSeqNum := 1 + lastSeqNum
+				skippedMessages := int64(msgSeqNum) - lastSeqNum
+
+				conn.logger.WithFields(bark.Fields{
+					"msgSeqNum":       msgSeqNum,
+					"expectedSeqNum":  expectedSeqNum,
+					"skippedMessages": skippedMessages,
+				}).Error("sequence number out of order")
 			}
 
-			hb.Beat()
-
-			switch rmc.GetType() {
-			case store.ReadMessageContentType_MESSAGE:
-
-				msg := rmc.GetMessage()
-				msgSeqNum := common.SequenceNumber(msg.Message.GetSequenceNumber())
-
-				// XXX: Sequence number check to make sure we get monotonically increasing
-				// sequence number.
-				// We just log and move forward
-				// XXX: Note we skip the first message check here because we can start from
-				// a bigger sequence number in case of restarts
-				if conn.extCache.destType != shared.DestinationType_TIMER {
-					if lastSeqNum+1 != int64(msgSeqNum) {
-						// FIXME: add metric to help alert this case
-						expectedSeqNum := 1 + lastSeqNum
-						skippedMessages := int64(msgSeqNum) - lastSeqNum
-
-						conn.logger.WithFields(bark.Fields{
-							"msgSeqNum":       msgSeqNum,
-							"expectedSeqNum":  expectedSeqNum,
-							"skippedMessages": skippedMessages,
-						}).Error("sequence number out of order")
-					}
-				} else {
-					// T471157 For timers, do not signal discontinuities to ack manager, since discontinuities are frequent
-					msgSeqNum = 0
-				}
+			// update the lastSeqNum to this value
+			lastSeqNum = msg.Message.GetSequenceNumber()
 
-				// update the lastSeqNum to this value
-				lastSeqNum = msg.Message.GetSequenceNumber()
+			// convert this to an outMessage
+			cMsg := cherami.NewConsumerMessage()
+			cMsg.EnqueueTimeUtc = msg.Message.EnqueueTimeUtc
+			cMsg.Payload = msg.Message.Payload
 
-				// convert this to an outMessage
-				cMsg := cherami.NewConsumerMessage()
-				cMsg.EnqueueTimeUtc = msg.Message.EnqueueTimeUtc
-				cMsg.Payload = msg.Message.Payload
+			cMsg.AckId = common.StringPtr(conn.extCache.ackMgr.getNextAckID(storeHostAddress(msgAddr), msgSeqNum))
 
-				cMsg.AckId = common.StringPtr(conn.extCache.ackMgr.getNextAckID(storeHostAddress(msg.GetAddress()), msgSeqNum))
-				// write the message to the msgsCh so that it can be delivered
-				// after being stored on the cache.
-				// 1. either there are no listeners
-				// 2. the buffer is full
-				// Wait until the there are some listeners or we are shutting down
+			if conn.extCache.destType != shared.DestinationType_TIMER {
+
+				enqueueTime := common.UnixNanoTime(msg.Message.GetEnqueueTimeUtc())
+				now := common.Now()
+
+				// check if the messages have already 'expired'; ie, if it falls outside the
+				// skip-older window. ignore (ie, don't skip any messages), if skip-older is '0'.
+				if skipOlderNanos > 0 && enqueueTime < (now-skipOlderNanos) {
+					continue loop
```

Comment:
no, we should not have to. since we have dropped the message on the floor (ie, not sent it out) we should not increment 'localMsgs' and therefore not have to renew credits, etc.